File: problems/0003_longest-substring-without-repeating-characters/index-of/solution.go
func lengthOfLongestSubstring(s string) int {
    max, l := 0, len(s)
    for start, end := 0, 0; end < l; {
        idx := indexOf(s, s[end], start, end)
        end += 1
        if idx == -1 {
            m := end - start
            if m > max {
                max = m
            }
        } else {
            start = idx + 1
        }
    }
    return max
}

func indexOf(s string, c byte, start, end int) int {
    for i := start; i < end; i += 1 {
        if s[i] == c {
            return i
        }
    }
    return -1
}

File: problems/0003_longest-substring-without-repeating-characters/index-of/solution.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    'use strict'
    function indexOf(c, start, end) {
        for (let i = start; i < end; ++i)
            if (s[i] === c)
                return i
    }
    let max = 0
    for (let start = 0, end = 0; end < s.length;) {
        const idx = indexOf(s[end], start, end++)
        if (idx === undefined)
            max = Math.max(max, end - start)
        else
            start = idx + 1
    }
    return max
};
